healsparse/healSparseMap.py

```diff
@@ -47,12 +47,83 @@ def read(cls, filename, nsideCoverage=None, pixels=None):
             return cls(healpixMap=healpixMap, nsideCoverage=nsideCoverage, nest=True)
         elif 'PIXTYPE' in hdr and hdr['PIXTYPE'].rstrip() == 'HEALSPARSE':
             # This is a sparse map type.  Just use fits for now.
+            covIndexMap, sparseMap, nsideSparse = cls._readHealSparseFile(filename, pixels=pixels)
+            return cls(covIndexMap=covIndexMap, sparseMap=sparseMap, nsideSparse=nsideSparse)
+        else:
+            raise RuntimeError("Filename %s not in healpix or healsparse format." % (filename))
+
+    @staticmethod
+    def _readHealSparseFile(filename, pixels=None):
+        """
+        Read a healsparse file, optionally with a set of coverage pixels
+        """
+        covIndexMap = fitsio.read(filename, ext='COV')
 
-            covIndexMap = fitsio.read(filename, ext='COV')
+        if pixels is None:
+            # Read the full map
             sparseMap, sHdr = fitsio.read(filename, ext='SPARSE', header=True)
-            return cls(covIndexMap=covIndexMap, sparseMap=sparseMap, nsideSparse=sHdr['NSIDE'])
+            nsideSparse = sHdr['NSIDE']
         else:
-            raise RuntimeError("Filename %s not in healpix or healsparse format." % (filename))
+            if len(np.unique(pixels)) < len(pixels):
+                raise RuntimeError("Input list of pixels must be unique.")
+
+            # Read a partial map
+            #fits = fitsio.FITS(filename)
+            with fitsio.FITS(filename) as fits:
+
+                hdu = fits['SPARSE']
+                sHdr = hdu.read_header()
+
+                nsideSparse = sHdr['NSIDE']
+                nsideCoverage = hp.npix2nside(covIndexMap.size)
+
+                bitShift = 2 * int(np.round(np.log(nsideSparse / nsideCoverage) / np.log(2)))
+                nFinePerCov = 2**bitShift
+
+                imageType = False
+                if hdu.get_exttype() == 'IMAGE_HDU':
+                    # This is an image extension
+                    # Currently, this is the only supported type.
+                    sparseMapSize = hdu.get_dims()[0]
+                    imageType = True
+                else:
+                    # This is a table extension
+                    raise RuntimeError("Table extension not yet supported.")
+                    sparseMapSize = hdu.get_nrows()
+
+                nCovPix = sparseMapSize / nFinePerCov - 1
+                covPix, = np.where((sparseMapSize - 2**bitShift) !=
+                                   (np.arange(hp.nside2npix(nsideCoverage), dtype=np.int64) * 2**bitShift) + covIndexMap)
+
+                # Find which pixels are in the coverage map
+                sub = np.clip(np.searchsorted(covPix, pixels), 0, covPix.size - 1)
```

Comment:
This is very clever! I don't know if you could just have used `ok  = np.in1d(covPix, pixels)` instead and `sub = covPix[ok]` or is this faster?

Reply:
Quite frankly, I didn't know about `np.in1d`.  (There are so many different ways to do things in numpy...)  And in spite of the call sequence being definitely cleaner, it's way slower than the `searchsorted` method.  Huh.